src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Wasi.cs

```diff
@@ -0,0 +1,58 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Sockets
+{
+    internal sealed partial class SocketAsyncEngine : IThreadPoolWorkItem
+    {
+        private async Task EventLoop()
+        {
+            try
+            {
+                SocketEventHandler handler = new SocketEventHandler(this);
+                while (true)
+                {
+                    int numEvents = EventBufferCount;
+                    await WaitForAnySocketPollable().ConfigureAwait(false);
+
+                    // The native shim is responsible for ensuring this condition.
+                    Debug.Assert(numEvents > 0, $"Unexpected numEvents: {numEvents}");
+
+                    // Only enqueue a work item if the stage is NotScheduled.
+                    // Otherwise there must be a work item already queued or another thread already handling parallelization.
+                    if (handler.HandleSocketEvents(numEvents) &&
+                        Interlocked.Exchange(
+                            ref _eventQueueProcessingStage,
+                            EventQueueProcessingStage.Scheduled) == EventQueueProcessingStage.NotScheduled)
+                    {
+                        ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Environment.FailFast("Exception thrown from SocketAsyncEngine event loop: " + e.ToString(), e);
+            }
+        }
+
+#pragma warning disable CA1822 // TODO remove
+        private Task WaitForAnySocketPollable()
+        {
+            TaskCompletionSource todo = new TaskCompletionSource();
+            /* TODO something like this, but with handle->pollable conversion
```

Comment:
libc's public interface only gives you access to posix-style file descriptors. 
Internally, it has a table that maps those descriptors to and from WASIp2 resources. 
That table isn't public, but you can cheat and access it anyway by copying 
https://github.com/WebAssembly/wasi-libc/blob/230d4be6c54bec93181050f9e25c87150506bdd0/libc-bottom-half/headers/private/wasi/descriptor_table.h and using those APIs. 
See https://github.com/dicej/mio/blob/a1853bb4882a85f5c43fa95d1e08235581d239b8/src/sys/wasi/mod.rs#L121-L285 for a Rust example where we use that table.